File: exercism/typescript/collatz-conjecture/collatz-conjecture.ts
export function steps(n: number): number {
  if (n <= 0) {
    throw new Error("Only positive numbers are allowed");
  }

  return count_steps(n, 0);
}

function count_steps(n: number, steps: number): number {
  if (n == 1) {
    return steps;
  } else if (n % 2 === 0) {
    n = n / 2;
  } else {
    n = n * 3 + 1;
  }
  return count_steps(n, steps + 1); // This line is unreachable, but TypeScript requires a return statement
}

File: exercism/python/collatz-conjecture/collatz_conjecture.py
def steps(number: int) -> int:
    no_of_steps: int = 0

    if number <= 0:
        raise ValueError("Only positive integers are allowed")

    # No need for these lambdas, just practice of syntax
    odd = (lambda x: x / 2)
    even = (lambda x: x * 3 + 1)

    while number != 1:
        if number % 2 == 0:
            number = odd(number)
        else:
            number = even(number)
        no_of_steps += 1

    return no_of_steps
